apps/tangle-dapp/containers/ManageProfileModalContainer/IndependentAllocationStep.tsx

```diff
@@ -0,0 +1,331 @@
+  FC,
+  SetStateAction,
+  useCallback,
+  useMemo,
+  useState,
+} from 'react';
+import { Cell, Pie, PieChart, Tooltip as RechartsTooltip } from 'recharts';
+import { z } from 'zod';
+
+import BnChartTooltip from '../../components/BnChartTooltip';
+import { ChartColor, TANGLE_TOKEN_UNIT } from '../../constants';
+import useStakingLedgerRx from '../../hooks/useStakingLedgerRx';
+import { ServiceType } from '../../types';
+import { formatTokenBalance } from '../../utils/polkadot';
+import AllocationInput from './AllocationInput';
+import { RestakingAllocationMap } from './types';
+
+type EntryName = 'Remaining' | ServiceType;
+
+type AllocationDataEntry = {
+  name: EntryName;
+  value: number;
+};
+
+export type IndependentAllocationStepProps = {
+  allocations: RestakingAllocationMap;
+  setAllocations: Dispatch<SetStateAction<RestakingAllocationMap>>;
+};
+
+function getPercentageOfTotal(amount: BN, total: BN): number {
+  // Avoid division by zero.
+  if (total.isZero()) {
+    throw new Error('Total should not be zero');
+  }
+
+  assert(amount.lte(total), 'Amount should be less than or equal to total');
+
+  // It's safe to convert to a number here, since the
+  // value will always be fraction between 0 and 1.
+  return amount.mul(new BN(100)).div(total).toNumber() / 100;
+}
+
+function getServiceChartColor(service: ServiceType): ChartColor {
+  switch (service) {
+    case ServiceType.ZK_SAAS_MARLIN:
+    case ServiceType.ZK_SAAS_GROTH16:
+      return ChartColor.Blue;
+    case ServiceType.DKG_TSS_CGGMP:
+      return ChartColor.Lavender;
+    case ServiceType.TX_RELAY:
+      return ChartColor.Green;
+  }
+}
+
+function getChartColor(entryName: EntryName): ChartColor {
+  switch (entryName) {
+    case 'Remaining':
+      return ChartColor.DarkGray;
+    default:
+      return getServiceChartColor(entryName);
+  }
+}
+
+export function cleanAllocations(
+  allocations: RestakingAllocationMap
+): [ServiceType, BN][] {
+  return Object.entries(allocations)
+    .filter(([_service, amount]) => amount !== null)
+    .map(([serviceString, amount]) => {
+      const service = z.nativeEnum(ServiceType).parse(serviceString);
+
+      assert(
+        amount !== null,
+        'Entries without amounts should have been filtered out'
+      );
+
+      return [service, amount];
+    });
+}
+
+const IndependentAllocationStep: FC<IndependentAllocationStepProps> = ({
+  allocations,
+  setAllocations,
+}) => {
+  const { data: stakedBalance } = useStakingLedgerRx(
+    useCallback((ledger) => ledger.total.toBn(), [])
+  );
+
+  // Max restaking amount = 50% of the total staked balance, which
+  // is equivalent to dividing the staked balance by 2. This is taken
+  // from Tangle's source code, as it seems that it is not obtainable
+  // from the Polkadot API.
+  // See: https://github.com/webb-tools/tangle/blob/8be20aa02a764422e1fd0ba30bc70b99d5f66887/runtime/mainnet/src/lib.rs#L1137
+  const maxRestakingAmount = stakedBalance?.divn(2) ?? null;
```

Comment:
adding `useMemo` and `useCallback` in below functions and variables (from line 103 to 203)

Reply:
I've added usage of `useMemo` and `useCallback` as suggested. Do note that some of the variables/computations are simple operations and did not require `useMemo`.